### PR TITLE
[WIP] refac DE

### PR DESCRIPTION
**Pull request by PierreBoyeau**
What I add in mind to refactor #342 based on Romain's work on Importance Sampling
There is still a couple things I need to do so that Importance sampling works.

I think adding tests specifically for importance sampling might be interesting.

### CODE REVIEW

File: scvi/inference/posterior.py [hunk 1]
```diff
@@ -224,62 +224,103 @@ def sample_scale_from_batch(self, n_samples, batchid=None, selection=None):
...
+            Remember that computing the Bayes Factor requires sampling
+            q(z_A | x_A) and q(z_B | x_B)
+
+        :param idx1: MASK of subpopulation cells 1
+        :param idx2: MASK of subpopulation cells 2
+        :param batchid1: batch ids of cells 1
+        :param batchid2: batch ids of cells 2
+        :param genes: Names of genes for which Bayes factors will be computed
+        :param n_samples: Number of times the posterior will be sampled for each pop
+        :param M_permutation: Number of times we will "mix" posterior samples in step 2
```
Comment (romain-lopez) [inline scvi/inference/posterior.py]: can we write something more precise ? I'm open to changing the name M_permutation. Also, M_permutation should follow sample_pairs ? Ideally first sample_pairs and then M_permutation since the latter is useful only provided that the former is useful only if the latter is True. for example, could we call M_permutation n_pairs ?

File: scvi/inference/posterior.py [hunk 2]
```diff
@@ -224,62 +224,103 @@ def sample_scale_from_batch(self, n_samples, batchid=None, selection=None):
...
+        :param all_stats: Returns either Bayes factors of complete info
```
Comment (romain-lopez) [inline scvi/inference/posterior.py]: Typo and what is complete info ?

File: scvi/inference/posterior.py [hunk 3]
```diff
@@ -224,62 +224,103 @@ def sample_scale_from_batch(self, n_samples, batchid=None, selection=None):
...
+        :param sample_pairs: Activates step 2 described above
```
Comment (romain-lopez) [inline scvi/inference/posterior.py]: from my understanding, step 2 is always active.  Just you either subsample a certain number of pairs or you don't.

File: scvi/inference/posterior.py [hunk 4]
```diff
@@ -224,62 +224,103 @@ def sample_scale_from_batch(self, n_samples, batchid=None, selection=None):
...
+        Computes gene specfic Bayes factors using masks idx1 and idx2
+
+        To that purpose we sample the Posterior in the following way:
+            1. The posterior is sampled n_samples times for each subpopulation
+            2. For computation efficiency (posterior sampling is quite expensive), instead of
+            comparing element-wise the obtained samples, we can permute posterior samples.
+            Remember that computing the Bayes Factor requires sampling
+            q(z_A | x_A) and q(z_B | x_B)
+
+        :param idx1: MASK of subpopulation cells 1
```
Comment (romain-lopez) [inline scvi/inference/posterior.py]: mask = binary array of size n_cells in the posterior ? You need to expand a bit on that since it's unusual

File: scvi/inference/posterior.py [hunk 5]
```diff
@@ -224,62 +224,103 @@ def sample_scale_from_batch(self, n_samples, batchid=None, selection=None):
...
+        :param idx2: MASK of subpopulation cells 2
+        :param batchid1: batch ids of cells 1
```
Comment (romain-lopez) [inline scvi/inference/posterior.py]: We might have to expand a little bit on that. That means that each cell from batch 1 is going to be reimputed in all the different batches of the LIST batchid1. Chenling can explain to you why this is useful in multiple batches studies. This is not the batch_id of every cell in the list idx1.

File: scvi/inference/posterior.py [hunk 6]
```diff
@@ -224,62 +224,103 @@ def sample_scale_from_batch(self, n_samples, batchid=None, selection=None):
...
+            return pd.Series(data=bayes1, index=self.gene_dataset.gene_names)
 
     @torch.no_grad()
     def one_vs_all_degenes(self, subset=None, cell_labels=None, min_cells=10,
-                           n_samples=None, M_permutation=None, output_file=False,
+                           n_samples=1000, M_permutation=None, output_file=False,
                            save_dir='./', filename='one2all'):
+        """
+
+        :param subset: MASK: subset of cells you are interested in
```
Comment (romain-lopez) [inline scvi/inference/posterior.py]: Define MASK more precisely

File: scvi/inference/posterior.py [hunk 7]
```diff
@@ -224,62 +224,103 @@ def sample_scale_from_batch(self, n_samples, batchid=None, selection=None):
...
+        :param cell_labels: Labels of cells
```
Comment (romain-lopez) [inline scvi/inference/posterior.py]: idem, is this a list, of what ?

File: scvi/inference/posterior.py [hunk 8]
```diff
@@ -224,62 +224,103 @@ def sample_scale_from_batch(self, n_samples, batchid=None, selection=None):
...
+        :param min_cells: Ceil number of cells used to compute Bayes Factors
+        :param n_samples: see `differential_expression_score`
```
Comment (romain-lopez) [inline scvi/inference/posterior.py]: You might have to copy over ? Is it good practice to see "go to xxx" ?

Comment (romain-lopez) [inline scvi/inference/posterior.py]: You might have to copy over ? Is it good practice to see "go to xxx" ?

File: scvi/inference/posterior.py [hunk 9]
```diff
@@ -643,47 +705,149 @@ def entropy(hist_data):
...
         # then constitutes the pairs
-        first_set = samples[u]
-        second_set = samples[v]
+        first_samples = samples[u]
+        second_samples = samples[v]
+        first_weights = weights[u]
+        second_weights = weights[v]
+        to_sum = first_weights[:, np.newaxis] * second_weights[:, np.newaxis] * (
+                first_samples >= second_samples)
+        incomplete_weights = first_weights * second_weights
```
Comment (romain-lopez) [inline scvi/inference/posterior.py]: I had some problem in practise with the imcomplete weights which would sometimes sum to zero. So it seems like using importance sampling means subselecting pairs of cells and using all the importance samples to compute the Bayes factor. I tried here to subsample from both the cells and the importances samples weight but it is really annoying in practise.

**Comment by romain-lopez** [comment]
- If you want to use importance sampling, we might have to push some changes in the VAE code to support for importance weight calculations (see scANVI repo)
- I also propose in the scANVI repo a much nicer way to reimputing gene expression on another batch than creating another data loader (which sounds quite extreme and dangerous to me). It consists in modifying the inference function in VAEs.
From here 
https://github.com/YosefLab/scVI/blob/master/scvi/inference/posterior.py#L205
to there 
https://github.com/chenlingantelope/HarmonizationSCANVI/blob/master/scvi/models/vae.py#L142

File: scvi/inference/posterior.py (hunk 8 above)
Comment (jeff-regier) [reply scvi/inference/posterior.py]: Better to copy
